### ISSUE TITLE
BASiNET

### ISSUE THREAD

**Issue by EricIto** [Bioconductor/Contributions#547]
- Repository: https://github.com/EricIto/BASiNET

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by mtmorgan** [comment]
Use Biostrings rather than seqinr. Your vignette must be a full exposition of package use, with fully evaluated R code chunks. Make these changes and submit a brief note indicating what you have done. Your package will then be considered for review.

**Comment by EricIto** [comment]
I've changed the package as you advised, now I'm using Biostrings with the readBStringSet () function. And I also changed the vignette, now I have detailed more like that using function and parameters

**Comment by EricIto** [comment]
Dear Bioconductor, Dear @mtmorgan,

I have three warning/error :

- checking whether package BASiNET can be installed ... [7s/5s] WARNING
Found the following significant warnings:
  Warning: 'rgl_init' failed, running with rgl.useNULL = TRUE

- Warning in rgl.init(initValue, onlyNULL) :
  RGL: unable to open X11 display
Warning: 'rgl_init' failed, running with rgl.useNULL = TRUE

- ERROR: dependencies RWeka, rmcfs are not available for package BASiNET

 Any suggestions on what I can do ? I tried to fix these errors but the problem is that on my computer these errors do not happen

**Comment by dvantwisk** [comment]
Both problems with the error and warnings are on our end.  More specifically, CRAN binaries are not available for R 3.5 on the Mac yet, so this error cannot be corrected until the binaries are available.  You can ignore the merida2 build error for now.  You can also ignore the warning on malbec2 involving rgl.  This is a warning we are still working in resolve on our end.  We will start reviewing your package now.

**Comment by dvantwisk** [comment]
I've been reviewing your package and I'd just like you to correct this issue before I continue.

vignettes
- [REQUIRED] As it stands now, the only runnable code in your vignette are some
knittr commands.  Please put your variables and `classfication()` call into a
code block so that this sections can be run.

**Comment by EricIto** [comment]
I put the examples in a block of code. See if I did it correctly please

**Comment by dvantwisk** [comment]
You need to write a code block as follows:
\```{r}

\```
You can run the command `R CMD Stangle <vignette_file>` to create an R file that shows all runnable code in the vignette.  You can use this to check whether you got it right.

Before I give you the rest of the review, I do need to confirm that you know that `BASiNET` is already a package on CRAN.  Once your package is accepted here, you must remove the package from CRAN since having the same packages on two concurrent repositories can cause problems.  Do you accept this?

**Comment by EricIto** [comment]
I agree, when my package is accepted in BASiNET I will remove from CRAN

**Comment by dvantwisk** [comment]
Here is a review to get you started.  There may be more things I need to bring up after you make these changes.  I'm having trouble with all the matrices you are creating in `classification.R:65-136`.  This comes off to me as an odd way to organize this data, but as it stands, most of the functions that help `classification()` that you have created are not vectorized and changing this method of organization may be a rather large overhaul.  I'm still deciding what to do for this part.  In the meantime, please make the following changes:

GENERAL:
- [REQUIRED] Add `.DS_Store` to `.gitignore`
- [REQUIRED] In your Roxygen headers be sure to be specific about what type of
argument a parameter required (e.g. `numeric()`, `logical(1)`, `data.frame`).  Also
specify specific exceptable files types, etc.
- [CONSIDER] `return()` at end of functions generally not needed.

DESCRIPTION:
- [REQUIRED] update required R version to 3.5.0

R/classification.R
- [REQUIRED] Use `match.arg()` to improve style of function declaration.
For example, the parameter `interval.type` can have two values "J48" and "RF".
The parameter in the function declaration can be written
as `interval.type=c('J48', 'RF')`.  The value that the user
uses can be obtained with `classifier <- match.arg(classifier)`.  The
first value (in this case `J48`) is the default argument.  Using this method
makes it clear to users what the allowed inputs to a function are and may also
make error catching easier.
- [REQUIRED] L35: For parameters with a default value you can use e.g. `step = 1` 
inside the function argument instead of calling `missing()` to set values.
- [REQUIRED] L76, and many other places: use `seq_along()` or `seq_len()`
instead of `1:...`.  This avoids the potentially unexpected case where `1:0`
returns `1 0` and in addition is also slightly faster.
- [REQUIRED] Use `message()` instead of `print()` for diagnostic messages.
- [REQUIRED] L64-140: This block looks very inefficient.  In would be far better
to not iteratively add on columns to the matrices and instead find a way to
vectorize `measures()`.  It may be better to create seperate matrices for mRNA,
lncRNA, and srcRNA instead of iterating across `numClass` and using the long
if-else statments in L78-97.
- [REQUIRED] L200-202: `data$CLASS[seq_len(seqMRNA)] <- 'mRNA'`

tests
- [CONSIDER] Adding tests to your package using the `testthat` package.

vignettes
- [REQUIRED] change name of vignette file to something informative as it will
have this name when searching for it using `vignette()`.

**Comment by EricIto** [comment]
About this:
[REQUIRED] L64-140: This block looks very inefficient. Would be better
iteratively add to columns to the matrices and instead find a way to
vectorize measures (). It may be better to create seperate matrices for mRNA,
lncRNA, and srcRNA instead of iterating across numClass and using the long
if-else statments in L78-97.

In this part of the code the measurement results are extracted and saved, these measurements are taken for each sequence, for each sequence there is an iteration of applying thresholds and extracting the measurements, this iteration is repeated N times and varies according to the network. That's why I'm adding columns in the matrix when I need it, and this feature that each network has a N of iterations is why I can not change that block of code

**Comment by dvantwisk** [comment]
At L64-140, I can't give a clear recommendation on how to change this code, so for the time being we can overlook it.  Please look into making the other changes though.

**Comment by dvantwisk** [comment]
Sorry for the wait.  I have some more things for you to look into.  These are mostly some improvements to  some of the issues I had in `classification.R`.  Please implement these changes and message me back when you have.  Also feel free to comment if you have any questions, etc.

GENERAL:
- [REQUIRED] add `.DS_STORE` files in every directory to `.gitignore`
- [CONSIDER] Removing commented out code.

R/classification.R
- We've been looking at the code in this file and have some recommendations to improve the look and runtime, but without compromising the structure of the package:
L74-90: Replace with:
```
if (k == 1) {
 	message("Analyzing mRNA from number: ")
        seq<-seqMRNA
} else if (k == 2) {
	message("Analyzing lncRNA from number: ")
	seq<-seqLNCRNA
} else {			# k == 3
	message("Analyzing sncRNA from number: ")
	seq<-seqSNCRNA
 }
```
This is a better way to write these conditionals.
L94: Remove `limitThreshold <- 0` (this is unneeded)
L100-127: Replace with:
```
			n_new_columns <- limitThreshold - ncol(averageShortestPathLengths)
			m_new <- matrix(nrow = numSeq, ncol = n_new_columns)
			averageShortestPathLengths <- cbind(averageShortestPathLengths, m_new)
			clusteringCoefficient <- cbind(clusteringCoefficient, m_new)
			...
			
			vector <- sapply(seq_len(limitThreshold), function(t) {
			    		net<<-threshold(t, net)
			    		measures(net)
			})
			
			cidx <- seq_len(nrow(vector))
			averageShortestPathLengths[aux + x, cidx] <- vector[1,]
			clusteringCoefficient[aux+x, cidx] <- vector[2,]
			...
```
The `...` here indicate where to continue writing code in the pattern we've specified.  This revised code avoid a for loop which included a cbind within the loop.  This operation is very expensive in R.
L131-154:
```
	numCol<-ncol(averageShortestPathLengths)
	
	listMatrix<-list(
	    averageShortestPathLengths,clusteringCoefficient,
	    standardDeviation,maximum,assortativity,betweenness,degree,
	    minimum,motifs3,motifs4)
	listMatrix <- lapply(
	    listMatrix, reschedule,
	    length(seqMRNA),length(seqLNCRNA),length(seqSNCRNA)
	)
	
	namesMeasure<-c(
	    "Average shortest path length", "Cluster Coefficient",
	    "Standard deviation", "Maximum", "Assortativity",
	    "Betweenness", "Degree", "Minimal", "Motifs 3", "Motifs 4"
	)
```
This just cleans up a section of code that repeated itself a lot.

R/threshold.R
L28-36:
```
       if(x!=1){
 		matriz<-as_adjacency_matrix(net)
 		matriz[(matriz==(x-1))]<-0
 		net<-graph_from_adjacency_matrix(matriz, mode="undirected")
 	}
 	return(net)
```
The first conditional is empty and is, therefore, not needed.

**Comment by dvantwisk** [comment]
Hello,

Have you been working on the changes that we requested?  We plan to have a new Bioconductor release in early May, so If you want your package to be included in that release, it will need to be accepted by mid-April.  Please message me back and feel free to ask for any help if it is needed.

**Comment by EricIto** [comment]
Yes, I've been working. I made the changes you suggested and I made other changes as well. Now I'm doing some tests and then I'm going to upload here

**Comment by dvantwisk** [comment]
Just messaging you back to check in on your progress.  This is another two week notice.  Do you want to keep this issue open?

**Comment by EricIto** [comment]
I have a doubt, in the checklist it speaks like this "The result of R CMD build must be less than 4MB". I made some changes in BASiNET, now it is possible to save a model and load this model so that it does not have to do the training of the data, but only the test of the sequences. I want to make available in the package some models for several species, however with that the size of the package was to 83.1mb. Do you have any suggestions for me?

**Comment by dvantwisk** [comment]
You'll need to submit the training data as an experiment data package.  We don't have any set guidelines for creating one, but I can give you an example of what the package should look like.  `alpineData` https://bioconductor.org/packages/release/data/experiment/html/alpineData.html is an experiment data package that contains large data files to be used as an example for the `alpine`  package https://bioconductor.org/packages/release/bioc/html/alpine.html.  You'll need to submit this package similarly to how you submitted this one, but leave a message stating `AdditionalPackage:https://github.com/EricIto/BASiNET` to show that the package is use in conjunction with another.  The closed issue for `alpineData` shows this https://github.com/Bioconductor/Contributions/issues/69.

**Comment by dvantwisk** [comment]
Hello,

I'm just sending out this message to those who are still working on their packages.  The package needs to be accepted by April 16th to be included in the upcoming Bioconductor 3.7 release.

**Comment by dvantwisk** [comment]
This issue has been inactive for a while.  It's been a while since anything has been pushed to it.  If the package is to be included in the 3.7 release, it must be accepted by April 16th, but that includes the time it takes us to review the changes.  Please make updates to this package soon or we will need to close this issue for the time being.

**Comment by dvantwisk** [comment]
I am going to close this issue for the time being since it has been inactive.  You may reopen this issue when you are ready to make further changes.
